### PR TITLE
Update Grand Potential Sintering model with strict mass conservation

### LINKED ISSUE #23329: Adding strict mass conservation option to GrandPotentialKernelAction
## Reason
Following issue  #22159, adding an option to use the strict mass conservation formulation of the Grand Potential Sintering model so it is simpler to switch between the original and the strict mass conservation version.

## Design
Add  a "mass_conservation" flag to the action kernel and a switch option for the chemical potential kernel creation, and add optional parameters to take concentration variable and material properties associated with the mass conservation formulation.

## Impact
Provide additional option for users to select the two formulations in Grand Potential Sintering model.

### PR DESCRIPTION
**Pull request by d-hermawan**
## Reason
The current GrandPotentialKernelAction sintering configuration for chemical potential evolution may result in mass loss due to discretization error.

## Design
Following the Grand Potential Phase Field formulation, the chemical potential equation is changed such that the relationship between concentration and chemical potential is specified at the expense of additional computational cost.

## Impact
- add additional properties in GrandPotentialSinteringMaterial associated with strict mass conservation
- add switch option in GrandPotentialKernelAction for strict mass conservation option
- add a test file and its gold file for the strict mass conservation option in GrandPotentialKernelAction under GrandPotentialPFM test

closes #23329

### CODE REVIEW

**Comment by d-hermawan** [comment]
@laagesen please review

**Comment by laagesen** [comment]
Thanks for submitting @d-hermawan - will try to get to it tomorrow but it may be Monday

File: modules/phase_field/src/action/GrandPotentialKernelAction.C [hunk 1]
```diff
@@ -244,50 +262,136 @@ GrandPotentialKernelAction::act()
...
+      } // for (unsigned int i = 0; i < n_c; ++i)
+      
+      // Chemical Potentials
+      for (unsigned int i = 0; i < n_w; ++i)
+      {
+        // coupling of c and w
+        InputParameters params = _factory.getValidParams("MatReaction");
+        params.set<NonlinearVariableName>("variable") = w_names[i];
+        params.set<std::vector<VariableName>>("v") = v4;
+        params.set<MaterialPropertyName>("mob_name") = mob_cw;
```
Comment (laagesen) [inline modules/phase_field/src/action/GrandPotentialKernelAction.C:364]: Here we just want `mob_name` to be a material property that is `-1` everywhere. Right now it is defined and calculated in `GrandPotentialSinteringMaterial`, but we create a lot of extra variables doing that. I can't find an example that would show that we could just define it as `-1` here, maybe @dschwen might have a suggestion?

File: modules/phase_field/include/materials/GrandPotentialSinteringMaterial.h [hunk 2]
```diff
@@ -111,6 +111,17 @@ class GrandPotentialSinteringMaterial : public DerivativeMaterialInterface<Mater
...
+  /// Body Force for mass conservation in conc and chempot coupling
+  MaterialProperty<Real> & _bodyforcev;
+  MaterialProperty<Real> & _bodyforces;
+
+  /// MatReaction Force for mass conservation in conc and chempot coupling
+  MaterialProperty<Real> & _matreactv;
+  MaterialProperty<Real> & _matreacts;
+  
+  /// mobility between c and w coupling
+  MaterialProperty<Real> & _mob_conc_w;
```
Comment (laagesen) [inline modules/phase_field/include/materials/GrandPotentialSinteringMaterial.h:123]: This is just a material property that is going to be `-1` everywhere, I am hoping we can define it right in the action rather than declaring it and computing here and passing in to the action. Please see further comments below

File: modules/phase_field/src/action/GrandPotentialKernelAction.C [hunk 3]
```diff
@@ -65,8 +66,20 @@ GrandPotentialKernelAction::validParams()
   parameters.addParam<bool>(
       "use_displaced_mesh", false, "Whether to use displaced mesh in the kernels");
   MultiMooseEnum anisotropy("true=1 false=0", "false");
+  parameters.addParam<bool>("mass_conservation", false, "Whether strict mass conservation formulation is used or not");
   parameters.addRequiredParam<MultiMooseEnum>(
       "anisotropic", anisotropy, "Whether or not each mobility is anisotropic");
+  parameters.addParam<std::vector<NonlinearVariableName>>(
+      "concentrations", "List of concentration variables for strict mass conservation");
+  parameters.addParam<std::vector<MaterialPropertyName>>(
+      "body_forces", "List of body forces coefficients for strict mass conservation formulation."
```
Comment (laagesen) [inline modules/phase_field/src/action/GrandPotentialKernelAction.C:75]: Please changes names as described above to `h_c_min` or something along those lines and also the description

File: modules/phase_field/src/action/GrandPotentialKernelAction.C [hunk 4]
```diff
@@ -65,8 +66,20 @@ GrandPotentialKernelAction::validParams()
...
+                     "Place in same order as switching_function_names.");
+  parameters.addParam<std::vector<MaterialPropertyName>>(
```
Comment (laagesen) [inline modules/phase_field/src/action/GrandPotentialKernelAction.C:77]: Please change names as described above to `h_over_kVa` or something like that

File: modules/phase_field/src/action/GrandPotentialKernelAction.C [hunk 5]
```diff
@@ -65,8 +66,20 @@ GrandPotentialKernelAction::validParams()
...
+      "mat_reacts", "List of mat react coefficients for strict mass conservation formulation."
+                     "Place in same order as switching_function_names.");
+  parameters.addParam<MaterialPropertyName>(
+      "mobility_cw", "Lcw", "Name of mobility to be used with concentration-chempot coupling."
```
Comment (laagesen) [inline modules/phase_field/src/action/GrandPotentialKernelAction.C:81]: Hopefully we can remove this an just create a material property for `-1` in the action itself

File: modules/phase_field/src/action/GrandPotentialKernelAction.C [hunk 6]
```diff
@@ -99,6 +112,11 @@ GrandPotentialKernelAction::act()
   auto implicity = getParam<bool>("implicit");
   auto displaced_mesh = getParam<bool>("use_displaced_mesh");
   auto aniso = getParam<MultiMooseEnum>("anisotropic");
+  const auto c_names = getParam<std::vector<NonlinearVariableName>>("concentrations");
+  const auto MRj = getParam<std::vector<MaterialPropertyName>>("mat_reacts");
```
Comment (laagesen) [inline modules/phase_field/src/action/GrandPotentialKernelAction.C:116]: Please change name

File: modules/phase_field/src/materials/GrandPotentialSinteringMaterial.C [hunk 7]
```diff
@@ -102,6 +102,11 @@ GrandPotentialSinteringMaterial::GrandPotentialSinteringMaterial(const InputPara
     _dkappa(declarePropertyDerivative<Real>("kappa", _phi_name)),
     _d2kappa(declarePropertyDerivative<Real>("kappa", _phi_name, _phi_name)),
     _gamma(declareProperty<Real>("gamma")),
+    _bodyforcev(declareProperty<Real>("BFv")),
```
Comment (laagesen) [inline modules/phase_field/src/materials/GrandPotentialSinteringMaterial.C:105]: Please change names

File: modules/phase_field/test/tests/GrandPotentialPFM/GrandPotentialStrictMassConservation.i [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+      free_energies_gr = 'omegav omegas'
+      additional_ops = 'phi'
+      gamma_grxop = gamma
+      mobility_name_op = Lv
+      kappa_op = kappa
+      free_energies_op = 'omegav omegas'
+      mass_conservation = 'true'
+      concentrations = 'c'
+      body_forces = 'BFv BFs'
+      mat_reacts = 'MRv MRs'
```
Comment (laagesen) [inline modules/phase_field/test/tests/GrandPotentialPFM/GrandPotentialStrictMassConservation.i:168]: Please change names

File: modules/phase_field/test/tests/GrandPotentialPFM/GrandPotentialStrictMassConservation.i [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+  nl_max_its = 20
+  l_max_its = 30
+
+  l_tol = 1e-8
+  nl_rel_tol = 1e-8
+  nl_abs_tol = 1e-8
+  
+  start_time = 0
+  dtmax = 0.05
+  num_steps = 10
```
Comment (laagesen) [inline modules/phase_field/test/tests/GrandPotentialPFM/GrandPotentialStrictMassConservation.i:217]: 2 or 3 time steps should be enough, unless you saw that longer than that was necessary for the non-mass conserved version to show violations of mass conservation

File: modules/phase_field/test/tests/GrandPotentialPFM/GrandPotentialStrictMassConservation.i [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+
+  [./TimeStepper]
```
Comment (laagesen) [inline modules/phase_field/test/tests/GrandPotentialPFM/GrandPotentialStrictMassConservation.i:219]: I would remove this and keep a fixed time step, just to keep the test input file as short as possible

File: modules/phase_field/test/tests/GrandPotentialPFM/GrandPotentialStrictMassConservation.i [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+  petsc_options_value = 'gmres      asm      ilu          1               31                 preonly'
+
+  automatic_scaling=true
+
+  nl_max_its = 20
+  l_max_its = 30
+
+  l_tol = 1e-8
+  nl_rel_tol = 1e-8
+  nl_abs_tol = 1e-8
```
Comment (laagesen) [inline modules/phase_field/test/tests/GrandPotentialPFM/GrandPotentialStrictMassConservation.i:213]: should not need this since you are not running to steady state

File: modules/phase_field/test/tests/GrandPotentialPFM/GrandPotentialStrictMassConservation.i [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+[]
+
+[Preconditioning]
+  [./SMP]
+    type = SMP
+    full = true
+  [../]
+[]
+
+[Debug]
```
Comment (laagesen) [inline modules/phase_field/test/tests/GrandPotentialPFM/GrandPotentialStrictMassConservation.i:193]: Please remove to keep test input file short

File: modules/phase_field/test/tests/GrandPotentialPFM/GrandPotentialStrictMassConservation.i [hunk 13]
```diff
@@ -0,0 +1,230 @@
...
+    x1 = 170
+    y1 = 270
+    z1 = 0
+    radius = 100
+    invalue = 1
+    outvalue = 0
+  [../]
+[]
+
+[Functions]
```
Comment (laagesen) [inline modules/phase_field/test/tests/GrandPotentialPFM/GrandPotentialStrictMassConservation.i:81]: you can set T = 1600 in the initial conditions for the AuxVariable so you can remove this to keep the input file short

File: modules/phase_field/test/tests/GrandPotentialPFM/GrandPotentialStrictMassConservation.i [hunk 14]
```diff
@@ -0,0 +1,230 @@
...
+  [./PolycrystalVariables]
+  [../]
+[]
+
+[AuxVariables]
+  [./bnds]
+  [../]
+  [./T]
+    order = CONSTANT
+    family = MONOMIAL
```
Comment (laagesen) [inline modules/phase_field/test/tests/GrandPotentialPFM/GrandPotentialStrictMassConservation.i:34]: can set `initial_condition = 1600` here

File: modules/phase_field/test/tests/GrandPotentialPFM/GrandPotentialStrictMassConservation.i [hunk 15]
```diff
@@ -0,0 +1,230 @@
...
+  []
+[]
+
+[AuxKernels]
+  [./bnds_aux]
+    type = BndsCalcAux
+    variable = bnds
+    execute_on = 'initial timestep_end'
+  [../]
+  [./T_aux]
```
Comment (laagesen) [inline modules/phase_field/test/tests/GrandPotentialPFM/GrandPotentialStrictMassConservation.i:179]: can remove this by setting T in initial conditions

File: modules/phase_field/src/action/GrandPotentialKernelAction.C [hunk 16]
```diff
@@ -244,50 +262,136 @@ GrandPotentialKernelAction::act()
...
+        params.set<std::vector<VariableName>>("v") = notarealvector;
+        params.set<std::vector<VariableName>>("args") = v0;
+        params.set<std::vector<MaterialPropertyName>>("Fj_names") = fj_temp;
+        params.set<std::vector<MaterialPropertyName>>("hj_names") = hj;
+        params.set<bool>("implicit") = implicity;
+        params.set<bool>("use_displaced_mesh") = displaced_mesh;
+        kernel_name = "Coupled_" + w_names[i] + "_" + all_etas[j];
+        _problem->addKernel("CoupledSwitchingTimeDerivative", kernel_name, params);
+      } //     for (unsigned int j = 0; j < n_etas + n_grs; ++j) coupledSwitchingTimeDeriv
+    } //for (unsigned int i = 0; i < n_w; ++i) chempot
```
Comment (laagesen) [inline modules/phase_field/src/action/GrandPotentialKernelAction.C:314]: It looks like there are some misaligned braces in a few places. This can be fixed by running the Clang-format utility on all your modified files. There are extensions to most text editors, such as VSCode or Atom, that will allow you to run this within the editor. The MOOSE test suite will also check for conformance with clang-format as part of the test suite and throw and error if it does not comply, it will give you the option to apply a patch to fix it, but it's easier just to install a plugin to your text editor and run prior to submitting

**Review by laagesen — CHANGES_REQUESTED** [review]
Please also update the documentation file `GrandPotentialKernelAction.md` to reflect the changes to the action

**Comment by laagesen** [comment]
Also please make sure your commit message ("Update Grand Potential Sintering model with strict mass conservation option") references the issue number 22329, so you will want to append #22329 to the end of the commit message. Also please make sure there are no trailing whitespaces at the end. I am not sure if you are currently able to see testing messages  from civet (https://civet.inl.gov/job/1330302/) so I am going to paste them here:

##########################################################################
ERROR: Your patch does not contain a valid ticket reference! (i.e. #1234)
Merge branch 'dhermawa' of https://github.com/d-hermawan/moose into test
Update Grand Potential Sintering model with strict mass conservation option

ERROR: The following files contain trailing whitespace after applying your patch:
	modules/phase_field/test/tests/GrandPotentialPFM/GrandPotentialStrictMassConservation.i

Run the "delete_trailing_whitespace.sh" script in your $MOOSE_DIR/scripts directory.
##########################################################################

File: modules/phase_field/src/action/GrandPotentialKernelAction.C [hunk 17]
```diff
@@ -15,10 +15,11 @@
 
 registerMooseAction("PhaseFieldApp", GrandPotentialKernelAction, "add_kernel");
 
+template <>
```
[the remark this answers is not in the document]
Comment (dschwen) [reply modules/phase_field/src/action/GrandPotentialKernelAction.C:18]: This should be a static member. The global function template approach for validParams has long been deprecated.

File: modules/phase_field/src/action/GrandPotentialKernelAction.C [hunk 18]
```diff
@@ -15,10 +15,11 @@
...
 InputParameters
-GrandPotentialKernelAction::validParams()
+validParams<GrandPotentialKernelAction>()
```
Comment (dschwen) [inline modules/phase_field/src/action/GrandPotentialKernelAction.C:20]: please undo these changes

File: modules/phase_field/src/action/GrandPotentialKernelAction.C [hunk 19]
```diff
@@ -99,6 +112,11 @@ GrandPotentialKernelAction::act()
   auto implicity = getParam<bool>("implicit");
   auto displaced_mesh = getParam<bool>("use_displaced_mesh");
   auto aniso = getParam<MultiMooseEnum>("anisotropic");
+  const auto c_names = getParam<std::vector<NonlinearVariableName>>("concentrations");
```
[the remark this answers is not in the document]
Comment (dschwen) [reply modules/phase_field/src/action/GrandPotentialKernelAction.C:115]: I believe this will always at least return an empty vector.

File: modules/phase_field/src/action/GrandPotentialKernelAction.C (hunk 16 above)
Comment (dschwen) [reply modules/phase_field/src/action/GrandPotentialKernelAction.C:314]: And please remove the commented out loop headers. Indentation should make this clear enough.

File: modules/phase_field/src/action/GrandPotentialKernelAction.C (hunk 1 above)
Comment (dschwen) [reply modules/phase_field/src/action/GrandPotentialKernelAction.C:364]: Just set it to `"-1"`

File: modules/phase_field/src/action/GrandPotentialKernelAction.C [hunk 20]
```diff
@@ -244,50 +262,136 @@ GrandPotentialKernelAction::act()
...
+          params.set<MaterialPropertyName>("mask") = BFj[j];
+          params.set<bool>("implicit") = implicity;
+          params.set<bool>("use_displaced_mesh") = displaced_mesh;
+          kernel_name = "MBD_" + w_names[i] + "_" + all_etas[j];
+          _problem->addKernel("MaskedBodyForce", kernel_name, params);
+        }
+      } // for (unsigned int i = 0; i < n_w; ++i)
+    } // case true
+  } // switch
 } // GrandPotentialKernelAction::act()
```
Comment (dschwen) [inline modules/phase_field/src/action/GrandPotentialKernelAction.C:397]: Please remove all these end-block comments. If the code is not readable enough  with just the correct indentation then the blocks span too many lines and the code should be refactored.

File: modules/phase_field/src/action/GrandPotentialKernelAction.C (hunk 16 above)
Comment (dschwen) [reply modules/phase_field/src/action/GrandPotentialKernelAction.C:314]: While you're at it, please also format the input files using the MOOSE VSCode/Atom plugin (or the `hit format file.i` command.

**Comment by dhermawa** [comment]
@laagesen please review

**Comment by laagesen** [comment]
@dhermawa glad you were able to get the changes pushed and thanks for making the updates! Before we proceed, can you please see the previous comment from @moosebuild about style changes? This is saying that the code still needs to be formatted to match the clang-format standards. You can do this either by applying the patch described in the latest @moosebuild comment, or by using the clang-format capability in VSCode as we talked about in our meeting the week before last.

Also the documentation will need to be updated, I don't see any updates to it in the latest push. Please update the documentation file GrandPotentialKernelAction.md to reflect the changes to the action. This file can be found at `moose/modules/phase_field/doc/content/source/action/GrandPotentialKernelAction.md`

**Comment by dhermawa** [comment]
@laagesen I've updated the documentation and apply the clang-format changes, please review.

**Review by laagesen — CHANGES_REQUESTED** [review]
@dhermawa running the test harness brought out a few issues I didn't notice first time around, sorry about that

**Comment by dhermawa** [comment]
@laagesen I've addressed the comments, please review

**Comment by dhermawa** [comment]
@laagesen please review

**Comment by dhermawa** [comment]
@laagesen I have rerun the gold file using "use_legacy_material_output" so it matches with PhaseFieldApp.C in moose base code, please activate the tests again

**Comment by dhermawa** [comment]
@laagesen I am unsure how to proceed with the docker error

**Comment by laagesen** [comment]
OK looks like we are good now! @loganharbour or @dschwen could you please merge?

File: modules/phase_field/src/actions/GrandPotentialKernelAction.C [hunk 21]
```diff
@@ -244,50 +261,133 @@ GrandPotentialKernelAction::act()
...
-    params.set<MaterialPropertyName>("f_name") = chis[i];
-    params.set<std::vector<VariableName>>("coupled_variables") = v0;
-    params.set<bool>("implicit") = implicity;
-    params.set<bool>("use_displaced_mesh") = displaced_mesh;
-    kernel_name = "ChiDt_" + w_names[i];
-    _problem->addKernel("SusceptibilityTimeDerivative", kernel_name, params);
+    std::vector<VariableName> v3;
+    v3.resize(n_w);
+    for (unsigned int i = 0; i < n_w; ++i)
+      v3[i] = w_names[i];
```
Comment (dschwen) [inline modules/phase_field/src/actions/GrandPotentialKernelAction.C:269]:
```suggestion
```

File: modules/phase_field/src/actions/GrandPotentialKernelAction.C [hunk 22]
```diff
@@ -244,50 +261,133 @@ GrandPotentialKernelAction::act()
...
-    params.set<bool>("use_displaced_mesh") = displaced_mesh;
-    params.set<MaterialPropertyName>("diffusivity") = M[i];
-    kernel_name = "MatDif_" + w_names[i];
-    if (aniso.get(i))
-      _problem->addKernel("MatAnisoDiffusion", kernel_name, params);
-    else
+      // MatDiffusion concentration (coupled with chempot)
+      params = _factory.getValidParams("MatDiffusion");
+      params.set<NonlinearVariableName>("variable") = c_names[i];
+      params.set<std::vector<VariableName>>("v") = v3;
```
Comment (dschwen) [inline modules/phase_field/src/actions/GrandPotentialKernelAction.C:289]:
```suggestion
      params.set<std::vector<VariableName>>("v") = w_names;
```

same for `v4`. This is just a suggestion. It seems to me that those vectors are not necessary. FYI, you could also just copy a vactor by assignment: `auto v3 = w_names;` . No need for the loops.

File: modules/phase_field/src/materials/GrandPotentialSinteringMaterial.C [hunk 23]
```diff
@@ -113,11 +119,16 @@ GrandPotentialSinteringMaterial::GrandPotentialSinteringMaterial(const InputPara
...
     _kappa_s(0.75 * _sigma_s * _int_width),
     _kappa_gb(0.75 * _sigma_gb * _int_width),
-    _kB(8.617343e-5) // eV/K
+    _kB(8.617343e-5), // eV/K
+    _mass_conservation(getParam<bool>("mass_conservation"))
 {
   if ((_switch > 1.0) || (_switch < 0.0))
     mooseError("GrandPotentialSinteringMaterial: surface_switch_value should be between 0 and 1");
 
+  if ((_mass_conservation) && (_solid_energy > 0))
```
Comment (dschwen) [inline modules/phase_field/src/materials/GrandPotentialSinteringMaterial.C:128]: FTI this would be fine, too (and preferred_

```suggestion
  if (_mass_conservation && _solid_energy > 0)
```

**Review by dschwen — DISMISSED** [review]
Looks fine to me. I leave it up to you if you want to make further changes.

File: modules/phase_field/src/actions/GrandPotentialKernelAction.C (hunk 21 above)
Comment (dhermawa) [reply modules/phase_field/src/actions/GrandPotentialKernelAction.C]: i tried implementing your suggestion, but it seems that because w_names is vector<NonlinearVariableName>, I cannot use it to set the kernel parameters that requires vector<VariableName>, so the assignment fails to compile.

**Comment by dhermawa** [comment]
@laagesen i have reverted back and recompile it to make sure it works. please activate the test

**Review by dschwen — APPROVED** [review]
Looks ok now. @laagesen please resolve your conversations of your comments are addressed.

**Review by laagesen — APPROVED** [review]

**Comment by laagesen** [comment]
Trying to figure out how to mark Outdated conversations as resolved (other than just deleting them)...

**Comment by loganharbour** [comment]
Larry doesn't have the ability to resolve. He's looked through and verified that they can be resolved so I'll resolve them

**Comment by loganharbour** [comment]
On to you for the magic button @dschwen
